src/legacy/core_plugins/data/public/filter/filter_manager/filter_manager.ts

```diff
@@ -0,0 +1,212 @@
+
+    // existing globalFilters should be mutated by appFilters
+    _.each(appFilters, function(filter, i) {
+      const match = _.find(globalFilters, function(globalFilter) {
+        return compareFilters(globalFilter, filter);
+      });
+
+      // no match, do nothing
+      if (!match) return;
+
+      // matching filter in globalState, update global and remove from appState
+      _.assign(match.meta, filter.meta);
+      appFilters.splice(i, 1);
+    });
+
+    return uniqFilters(appFilters.reverse().concat(globalFilters.reverse())).reverse();
+  }
+
+  private filtersUpdated(newFilters: Filter[]): boolean {
+    return !_.isEqual(this.filters, newFilters);
+  }
+
+  private static partitionFilters(filters: Filter[]): PartitionedFilters {
+    const [globalFilters, appFilters] = _.partition(filters, isFilterPinned);
+    return {
+      globalFilters,
+      appFilters,
+    };
+  }
+
+  private handleStateUpdate(newFilters: Filter[]) {
+    // This is where the angular update magic \ syncing diget happens
+    const filtersUpdated = this.filtersUpdated(newFilters);
+
+    // global filters should always be first
+    newFilters.sort(
+      (a: Filter, b: Filter): number => {
+        if (a.$state && a.$state.store === FilterStateStore.GLOBAL_STATE) {
+          return -1;
+        } else if (b.$state && b.$state.store === FilterStateStore.GLOBAL_STATE) {
+          return 1;
+        } else {
+          return 0;
+        }
+      }
+    );
+
+    this.filters = newFilters;
+    if (filtersUpdated) {
+      this.updated$.next();
+      // Fired together with updated$, because historically (~4 years ago) there was a fetch optimization, that didn't call fetch for very specific cases.
+      // This optimization seems irrelevant at the moment, but I didn't want to change the logic of all consumers.
+      this.fetch$.next();
+    }
+  }
+
+  /* Getters */
+
+  public getFilters() {
+    return this.filters;
+  }
+
+  public getAppFilters() {
+    const { appFilters } = this.getPartitionedFilters();
+    return appFilters;
+  }
+
+  public getGlobalFilters() {
+    const { globalFilters } = this.getPartitionedFilters();
+    return globalFilters;
+  }
+
+  public getPartitionedFilters(): PartitionedFilters {
+    return FilterManager.partitionFilters(this.filters);
+  }
+
+  public getUpdates$() {
+    return this.updated$.asObservable();
+  }
+
+  public getFetches$() {
+    return this.fetch$.asObservable();
+  }
+
+  /* Setters */
+
+  public async addFilters(filters: Filter[] | Filter, pinFilterStatus?: boolean) {
+    if (!Array.isArray(filters)) {
+      filters = [filters];
+    }
+
+    const { uiSettings } = npSetup.core;
+    if (pinFilterStatus === undefined) {
+      pinFilterStatus = uiSettings.get('filters:pinnedByDefault');
+    }
+
+    // set the store of all filters
+    // TODO: is this necessary?
+    const store = pinFilterStatus ? FilterStateStore.GLOBAL_STATE : FilterStateStore.APP_STATE;
```

Comment:
couldn't we get rid of this and setFilterStore method and rather just use `pinFilterStatus` in handleStateUpdate ?